### PR TITLE
Expose and test new GetBytes/GetByteCount overloads

### LINKED ISSUE #14255: Add Encoding.GetBytes(string, offset, count)
The Encoding class has methods for encoding the middle of a `char[]`, but not for the middle of a `string`.  A caller is forced to switch to unsafe (char*)  or to re-allocate as `char[]`.

```diff
    public abstract partial class Encoding : System.ICloneable
    {
...
        public unsafe virtual int GetByteCount(char* chars, int count) { throw null; }
        public virtual int GetByteCount(char[] chars) { throw null; }
        public abstract int GetByteCount(char[] chars, int index, int count);
        public virtual int GetByteCount(string s) { throw null; }
+       public int GetByteCount(string s, int index, int count) { throw null; }
...
        public unsafe virtual int GetBytes(char* chars, int charCount, byte* bytes, int byteCount) { throw null; }
        public virtual byte[] GetBytes(char[] chars) { throw null; }
        public virtual byte[] GetBytes(char[] chars, int index, int count) { throw null; }
        public abstract int GetBytes(char[] chars, int charIndex, int charCount, byte[] bytes, int byteIndex);
        public virtual byte[] GetBytes(string s) { throw null; }
+       public byte[] GetBytes(string s, int index, int count) { throw null; }
        public virtual int GetBytes(string s, int charIndex, int charCount, byte[] bytes, int byteIndex) { throw null; }
...
}
```

The `GetBytes(string, int, int)` method would allow for the typical "just give me a big enough array" while the `GetByteCount(string, int, int)` method would allow for ensuring that an existing buffer is sufficiently big to call the existing `GetBytes(string, int, int, byte[], int)` method.

### PR DESCRIPTION
**Pull request by hughbe**
This is currently a WIP - I'm having some problems exposing the API so want to see what the CI says...

Fixes #14255

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test Innerloop Ubuntu14.04 Release Build and Test please

File: src/System.Text.Encoding/tests/NegativeEncodingTests.cs [hunk 1]
```diff
@@ -39,18 +39,48 @@ public static unsafe void GetByteCount_Invalid(Encoding encoding)
         {
             // Chars is null
             Assert.Throws<ArgumentNullException>(encoding is ASCIIEncoding ? "chars" : "s", () => encoding.GetByteCount((string)null));
+#if netcoreapp
```
Comment (tarekgh) [inline src/System.Text.Encoding/tests/NegativeEncodingTests.cs]: in general I don't like to have many ifdef's in the code. we can wither move the cases that is netcoreapp specific to its own file or we can have test cases skipped for all platforms but netcoreapp. you may look at the PR https://github.com/dotnet/corefx/pull/16706 for similar idea.

**Comment by tarekgh** [comment]
@hughbe did you have a chance to look at the comment on this PR? thanks.

**Comment by hughbe** [comment]
Sorry for the delay. I've updated the PR. I won't be able to respond to any comments until Tuesday at least as I'm in the provinces for work, away from a laptop

**Comment by tarekgh** [comment]
thanks @hughbe LGTM.

**Pull request merged**